crates/kitsune_p2p/types/src/tx2/tx2_api.rs

```diff
@@ -0,0 +1,489 @@
+//! Usability api for tx2 kitsune transports.
+
+use crate::codec::*;
+use crate::tx2::tx2_frontend::*;
+use crate::tx2::tx2_utils::*;
+use crate::tx2::*;
+use crate::*;
+use futures::future::{FutureExt, TryFutureExt};
+use futures::stream::Stream;
+use std::collections::HashMap;
+use std::sync::atomic;
+
+static MSG_ID: atomic::AtomicU64 = atomic::AtomicU64::new(1);
+
+fn next_msg_id() -> u64 {
+    MSG_ID.fetch_add(1, atomic::Ordering::Relaxed)
+}
+
+type RSend<C> = tokio::sync::oneshot::Sender<KitsuneResult<C>>;
+type ShareRMap<C> = Arc<Share<RMap<C>>>;
+
+struct RMap<C: Codec + 'static + Send + Unpin>(HashMap<(ConHnd, u64), (RSend<C>, KitsuneTimeout)>);
+
+impl<C: Codec + 'static + Send + Unpin> RMap<C> {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    pub fn insert(&mut self, con: ConHnd, msg_id: u64, s_res: RSend<C>, timeout: KitsuneTimeout) {
+        self.0.insert((con, msg_id), (s_res, timeout));
+    }
+
+    pub fn respond(&mut self, con: ConHnd, msg_id: u64, c: C) {
+        if let Some((s_res, _)) = self.0.remove(&(con, msg_id)) {
+            let _ = s_res.send(Ok(c));
```

Comment:
Is the `let _` an error? Does it matter?

Reply:
It's normal that a receiver would be dropped when it times out and thus the respond would go nowhere - added a comment.